doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+```
+aci: (target = "cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (version 3.0; acl "Allow users to add entries grouping app passwords for a
+     particular app";
+     allow (add) 
+     userdn = "ldap:///all;)
+aci: (target = "ldap:///uid=($dn)-\*,cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (targetattr = "uid || displayName || userPassword || appName || entryUUID
+     || managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to add an app password for themselves";
+     allow (add) 
+     userdn = "ldap:///uid=($dn),cn=users,cn=accounts,$SUFFIX";)
+aci: (targetattr = "managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to search for their app passwords";
+     allow (search) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+aci: (targetattr = "displayName")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to read display names for their app
+     passwords";
+     allow (read) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+aci: (target = "ldap:///uid=\*,cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (targetattr = "uid || displayName || userPassword || appName || entryUUID
+     || managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to delete their app password";
+     allow (delete) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+```
+
+Backup and restore is out of scope of this extension, but it will probably work
+with the new functionality.
+
+I do not need to enforce anything for app passwords. I suppose it will then be
+best to implement some if-else branching in
+`daemons/ipa-slapi-plugins/ipa-pwd-extop/prepost.c`, where it will be tested if
+the Password attribute being modified is not from the subtree for app passwords,
+and if it is, then no check and enforcements, as well as no OTP validation will
+be performed for it.
+
+**New dependencies**: TBD, not expected for now
+
+## Feature management
+
+The login process will remain the same, only instead of entering the primary
+password, the user will enter an app password. And as already stated, the user
+then will not have access to the account management.
+
+### UI
+
+TBD
+
+### CLI
+
+| Command | Arguments | Options |
+| ------- | --------- | ------- |
+| apppw-add | username (user's uid), displayname, appname |  |
+| apppw-list | username (user's uid) |  |
+| apppw-del | username (user's uid), uuid |  |
+
+## Test Plan
+
+1. All existing tests must pass. This also ensures that the user can have only
+one password.
+2. Test app passwords:
+   1. Add one app password using WebUI and one using CLI.
+   2. Check that using these app passwords, it is possible to authenticate to
+      all services and systems, but not to the FreeIPA server itself for account
+      management.
+   3. Log in using the primary password.
```

Comment:
Log in to what? IPA or a system login?

Reply:
To FreeIPA. I rewrote the test plan to be more clear.

Reply:
And also elaborated it a bit more.